### PR TITLE
pbw_algbra: left_ideal eliminate

### PR DESCRIPTION
**Pull request by tthsqe12**

### CODE REVIEW

File: src/Rings/orderings.jl [hunk 1]
```diff
@@ -876,6 +877,70 @@ function is_mixed(M::MonomialOrdering)
...
+@doc Markdown.doc"""
+    is_elimination_ordering(M::MonomialOrdering, v::Vector{Int})
+    is_elimination_ordering(M::MonomialOrdering, v::Vector{<:MPolyElem})
+
+Return `true` if the given ordering eliminates the variables in `v`.
+"""
+function is_elimination_ordering(o::MonomialOrdering, sigmaC::Vector{Int})
+# This is probably correct: o is elimination ordering for sigma^C <=>
+#   For each j in sigma^C, the first non-zero weight M[i,j] in column j is >= 0,
+#   and the weights in columns corresponding to sigma in rows <= i are zero.
```
Comment (tthsqe12) [inline src/Rings/orderings.jl:1253]: @hannes14 Is this a correct characterization of elimination orderings by their weight matrices? Does the singular kernel have such a function somewhere?

**Comment by wdecker** [comment]
@tthsqe12 

`function eliminate(I::PBWAlgIdeal{-1, T, S}, sigmaC::Vector{Int}) where {T, S}
  R = base_ring(I)
  if !is_elimination_subalgebra_admissible(R, sigmaC)
    error("no elimination is possible: subalgebra is not admissible")
  end
  r = PolynomialRing(coefficient_ring(R), symbols(R))[1]
  o = monomial_ordering(r, Singular.ordering(base_ring(R.relations)))
  if is_elimination_ordering(o, sigmaC)
    z = _eliminate_via_given_ordering(I.sdata, sigmaC)
    return PBWAlgIdeal{-1, T, S}(R, z)`

How about adding another `else` here: write a function `is_admissible` for a monomial ordering w.r.t. to a given PBW-algebra and check whether the block ordering (dp, dp) corresponding to the given elimination situation is admissible?

`else
ok, w = _elimination_ordering_weights(base_ring(I), sigmaC)
    ok || error("could not find elimination ordering")`

Is your check implemented so that it decides whether an admissible elimination ordering exists? If so, please correct error message:  

`ok || error("An admissible elimination ordering does not exist")`

File: src/Rings/orderings.jl (hunk 1 above)
Comment (hannes14) [reply src/Rings/orderings.jl:1253]: one needs:
- M is of rank n (number of variables), i.e. "o" is an ordering
- first non-zero weight M[i,j] in each column j is >0, i.e. "o" is a well ordering
- elimination for a subset of variables: in the first row of M the entries corresponding to the to eliminated variables
  must be >0, all the other entries must be zero
Singular does not have a function to test "is_elimination_ordering", but can (via prepending a(v)) change every well ordering to an elimination ordering for these variables where a one is in v.

**Comment by tthsqe12** [comment]
One sticky question: Suppose `I` is an ideal of `K[x,y,z,w]`. Isn't the return of `eliminate(I, [x,y])` supposed to be an ideal of `K[z,w]`? (an not of `K[x,y,z,w]`)

**Comment by thofma** [comment]
> One sticky question: Suppose `I` is an ideal of `K[x,y,z,w]`. Isn't the return of `eliminate(I, [x,y])` supposed to be an ideal of `K[z,w]`? (an not of `K[x,y,z,w]`)

FWIW, Magma also stays inside the ring you start with.

**Comment by wdecker** [comment]
> FWIW, Magma also stays inside the ring you start with.

Dito for Singular and Macaulay2. Ideally, we should have two functions, say `eliminate` and `intersection_with_subring`,
where the latter implements what is done in the example below:

julia> R, (t, x, y, z) = PolynomialRing(QQ, ["t", "x", "y", "z"])
(Multivariate Polynomial Ring in t, x, y, z over Rational Field, fmpq_mpoly[t, x, y, z])

julia> I = ideal(R, [t-x, t^2-y, t^3-z])
ideal(t - x, t^2 - y, t^3 - z)

julia> A = [t]
1-element Vector{fmpq_mpoly}:
 t

julia> TC = eliminate(I, A)
ideal(-x*z + y^2, x*y - z, x^2 - y)

julia> base_ring(TC)
Multivariate Polynomial Ring in t, x, y, z over Rational Field

julia> S, (x, y, z) = PolynomialRing(QQ, ["x", "y", "z"]);

julia> F = hom(R, S, [0, x, y, z])
Map with following data
Domain:
=======
Multivariate Polynomial Ring in t, x, y, z over Rational Field
Codomain:
=========
Multivariate Polynomial Ring in x, y, z over Rational Field

julia> ITC = F(TC)
ideal(-x*z + y^2, x*y - z, x^2 - y)

julia> base_ring(ITC)
Multivariate Polynomial Ring in x, y, z over Rational Field

However, at this point, the OSCAR design differs from that of the other named systems: Defining `S` that way, `x, y, z` are lost for `R`. And, in a typical algebraic geometry session, I would need both rings `R` and `S` (for example, I might wish to project to other P2's as well). Is there a solution to this problem other than renaming the variables? If not, how should we carefully design `intersect_with_subrings`?

**Review by wdecker — APPROVED** [review]
Returning the elimination ideal in the given ring is o.k. for now. We will eventually create maps between PBW algebras at some point and can then show in the docu, how to create the elimination ideal in the corresponding subalgebra.

**Comment by tthsqe12** [comment]
@thofma any ideas on the nightly errors?

**Comment by thofma** [comment]
No clue, never seen this before. Maybe @fingolfin?

**Comment by tthsqe12** [comment]
nothingburger - the merge commit passed

**Pull request merged**